python/ray/util/ml_utils/checkpoint_manager.py

```diff
@@ -0,0 +1,433 @@
+            self._decide_what_to_do_with_checkpoint(persisted_checkpoint)
+
+        self._latest_checkpoint_id += 1
+
+    def _replace_latest_memory_checkpoint(self, memory_checkpoint: TrackedCheckpoint):
+        assert memory_checkpoint.storage_mode == CheckpointStorage.MEMORY
+        self._latest_memory_checkpoint = memory_checkpoint
+        # Avoid memory leaks on k8s pods
+        gc.collect()
+
+    def _replace_latest_persisted_checkpoint(
+        self, persisted_checkpoint: TrackedCheckpoint
+    ):
+        second_to_latest_persisted_checkpoint = self._latest_persisted_checkpoint
+        self._latest_persisted_checkpoint = persisted_checkpoint
+
+        if self._checkpoint_strategy.num_to_keep == 0:
+            self._maybe_delete_persisted_checkpoint(
+                second_to_latest_persisted_checkpoint
+            )
+
+    def _maybe_replace_best_persisted_checkpoint(
+        self, persisted_checkpoint: TrackedCheckpoint
+    ):
+        if self._best_persisted_checkpoint is None:
+            self._best_persisted_checkpoint = persisted_checkpoint
+        else:
+            old_score = self._get_checkpoint_score(self._best_persisted_checkpoint)
+            candidate_score = self._get_checkpoint_score(persisted_checkpoint)
+            if candidate_score >= old_score:
+                self._best_persisted_checkpoint = persisted_checkpoint
+
+    def _get_checkpoint_score(
+        self, checkpoint: TrackedCheckpoint
+    ) -> Tuple[bool, numbers.Number, int]:
+        checkpoint_score_attribute = (
+            self._checkpoint_strategy.checkpoint_score_attribute
+        )
+        if checkpoint_score_attribute not in checkpoint.metrics:
+            logger.error(
+                f"Result dict has no key: {checkpoint_score_attribute}. "
+                f"checkpoint_score_attr must be set to a key in the "
+                f"result dict. Valid keys are: {list(checkpoint.metrics.keys())}"
+            )
+            checkpoint_result = float("-inf")
+        else:
+            checkpoint_result = checkpoint.metrics[checkpoint_score_attribute]
+
+        checkpoint_score_order = self._checkpoint_strategy.checkpoint_score_order
+        if checkpoint_score_order == MAX:
+            order_factor = 1.0
+        else:
+            order_factor = -1.0
+
+        checkpoint_score = order_factor * checkpoint_result
+
+        if not isinstance(checkpoint_score, numbers.Number):
+            raise ValueError(
+                f"Unable to persist checkpoint for "
+                f"checkpoint_score_attribute: "
+                f"{checkpoint_score_attribute} with value "
+                f"{checkpoint_score}. "
+                f"This attribute must be numerical."
+            )
+
+        return (
+            not is_nan(checkpoint_score),
+            checkpoint_score if not is_nan(checkpoint_score) else 0,
+            checkpoint.id,
+        )
+
+    def _decide_what_to_do_with_checkpoint(self, checkpoint: TrackedCheckpoint):
```

Comment:
if this method only applies to persistent checkpoint, then rename to `_decide_what_to_do_with_persistent_checkpoint`?

Reply:
Renamed to `_process_persistent_checkpoint`